### PR TITLE
8302906: AArch64: Add SVE backend support for vector unsigned comparison

### PR DESCRIPTION
**Pull request by changpeng1997**
This patch implements unsigned vector comparison on SVE.

1: Test:
All vector API test cases[1][2] passed without new failure. Existing test cases can cover all unsigned comparison conditions for all kinds of vector.

2: Performance:
(1): Benchmark:
As existing benchmarks in panama repo (such as [3]) have some issues [4] (We will fix them in a separate patch.), I collected performance data with a reduced jmh benchmark [5]. e.g. for ByteVector unsigned comparison:

```
@Benchmark
    public void byteVectorUnsignedCompare() {
        for (int j = 0; j < 200; j++) {
            for (int i = 0; i < bspecies.length(); i++) {
                ByteVector av = ByteVector.fromArray(bspecies, ba, i);
                ByteVector ca = ByteVector.fromArray(bspecies, bb, i);
                av.compare(VectorOperators.UNSIGNED_GT, ca).intoArray(br, i);
            }
        }
    }
```

(2): Performance data

Before:

```
Benchmark                               Score(op/ms)     Error
ByteVector.UNSIGNED_GT#size(1024)       4.846            3.419
ShortVector.UNSIGNED_GE#size(1024)      3.055            1.369
IntVector.UNSIGNED_LT#size(1024)        3.475            1.269
LongVector.UNSIGNED_LE#size(1024)       4.515            1.812
```

After:

```
Benchmark                               Score(op/ms)     Error
ByteVector.UNSIGNED_GT#size(1024)       493.937          1.389
ShortVector.UNSIGNED_GE#size(1024)      5308.796         20.557
IntVector.UNSIGNED_LT#size(1024)        4944.744         10.606
LongVector.UNSIGNED_LE#size(1024)       8459.605         28.683
```

[1] https://github.com/openjdk/jdk/tree/master/test/jdk/jdk/incubator/vector
[2] https://github.com/openjdk/jdk/tree/master/test/hotspot/jtreg/compiler/vectorapi
[3] https://github.com/openjdk/panama-vector/blob/2aade73adeabdf6a924136b17fd96ccc95c1d160/test/micro/org/openjdk/bench/jdk/incubator/vector/operation/ByteMaxVector.java#L1459
[4] https://bugs.openjdk.org/browse/JDK-8282850
[5] https://gist.github.com/changpeng1997/d311127e1015c107197f9b56a92b0fae

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302906](https://bugs.openjdk.org/browse/JDK-8302906): AArch64: Add SVE backend support for vector unsigned comparison

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Eric Liu](https://openjdk.org/census#eliu) (@theRealELiu - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12725/head:pull/12725` \
`$ git checkout pull/12725`

Update a local copy of the PR: \
`$ git checkout pull/12725` \
`$ git pull https://git.openjdk.org/jdk pull/12725/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12725`

View PR using the GUI difftool: \
`$ git pr show -t 12725`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12725.diff">https://git.openjdk.org/jdk/pull/12725.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -1117,6 +1117,10 @@ void C2_MacroAssembler::sve_compare(PRegister pd, BasicType bt, PRegister pg,
     case BoolTest::ge: condition = Assembler::GE; break;
     case BoolTest::lt: z1 = zm; z2 = zn; condition = Assembler::GT; break;
     case BoolTest::gt: condition = Assembler::GT; break;
+    case BoolTest::ugt: condition = Assembler::HI; break;
+    case BoolTest::uge: condition = Assembler::HS; break;
+    case BoolTest::ult: z1 = zm; z2 = zn; condition = Assembler::HI; break;
+    case BoolTest::ule: z1 = zm; z2 = zn; condition = Assembler::HS; break;
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1123]: I'm sure you already know what to do here.

Comment (changpeng1997) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1123]: Ok, thanks. I will remove these switch-case statements.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 2]
```diff
@@ -3184,6 +3176,46 @@ template<typename R, typename... Rx>
...
+      default:
+        ShouldNotReachHere();
+        break;
+    }
+
+    f(0, 31), f((int)T & 1, 30), f((cond_op >> 1) & 1, 29);
+    f(0b01110, 28, 24), f(cond_op & 1, 23), f(T == T2D ? 1 : 0, 22);
+    f(1, 21), rf(Vm, 16), f(0b111001, 15, 10), rf(Vn, 5), rf(Vd, 0);
+  }
+
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3218]: This looks OK, but it's in the wrong place in the file. Look at C4.1 A64 instruction set encoding. These instructions are in the "Advanced SIMD three same" group, so they must appear in assembler_aarch64.hpp in the "Advanced SIMD three same" section.
This is the "AdvSIMD two-reg misc" section.

Comment (changpeng1997) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp]: Sorry for this error.

**Review by theRealAph — APPROVED** [review]
That's a lot cleaner. Thanks.

**Comment by changpeng1997** [comment]
> That's a lot cleaner. Thanks.

Thanks for your review.

**Review by theRealELiu — APPROVED** [review]
LGTM

**Pull request closed without merging**